packages/vscode-extension/src/commonlib/vscodeAzureSubscriptionProvider.ts

```diff
@@ -0,0 +1,263 @@
+      subscriptions.push({
+        authentication: authentication,
+        environment: environment,
+        credential: credential,
+        isCustomCloud: environment.isCustomCloud,
+        /* eslint-disable @typescript-eslint/no-non-null-assertion */
+        name: subscription.displayName!,
+        subscriptionId: subscription.subscriptionId!,
+        /* eslint-enable @typescript-eslint/no-non-null-assertion */
+        tenantId: tenantId,
+      });
+    }
+
+    return subscriptions;
+  }
+}
+
+export async function getSessionFromVSCode(
+  scopes?: string | string[],
+  tenantId?: string,
+  options?: vscode.AuthenticationGetSessionOptions
+): Promise<vscode.AuthenticationSession | undefined> {
+  return await vscode.authentication.getSession(Microsoft, getScopes(scopes, tenantId), options);
+}
+
+function ensureEndingSlash(value: string): string {
+  return value.endsWith("/") ? value : `${value}/`;
+}
+
+function getResourceScopes(scopes?: string | string[]): string[] {
+  if (scopes === undefined || scopes === "" || scopes.length === 0) {
+    scopes = ensureEndingSlash(getConfiguredAzureEnv().managementEndpointUrl);
+  }
+  return Array.from(new Set<string>(scopes));
+}
+
+function addTenantIdScope(scopes: string[], tenantId: string): string[] {
+  const scopeSet = new Set<string>(scopes);
+  scopeSet.add(`VSCODE_TENANT:${tenantId}`);
+  return Array.from(scopeSet);
+}
+
+function getScopes(scopes: string | string[] | undefined, tenantId?: string): string[] {
+  let scopeArr = getResourceScopes(scopes);
+  if (tenantId) {
+    scopeArr = addTenantIdScope(scopeArr, tenantId);
+  }
+  return scopeArr;
+}
+
+/**
+ * Represents a means of obtaining authentication data for an Azure subscription.
+ */
+export interface AzureAuthentication {
+  /**
+   * Gets a VS Code authentication session for an Azure subscription.
+   *
+   * @param scopes - The scopes for which the authentication is needed.
+   *
+   * @returns A VS Code authentication session or undefined, if none could be obtained.
+   */
+  getSession(scopes?: string[]): vscode.ProviderResult<vscode.AuthenticationSession>;
+}
+
+// These strings come from https://github.com/microsoft/vscode/blob/eac16e9b63a11885b538db3e0b533a02a2fb8143/extensions/microsoft-authentication/package.json#L40-L99
+const CustomCloudConfigurationSection = "microsoft-sovereign-cloud";
+const CloudEnvironmentSettingName = "environment";
+const CustomEnvironmentSettingName = "customEnvironment";
+
+enum CloudEnvironmentSettingValue {
+  ChinaCloud = "ChinaCloud",
+  USGovernment = "USGovernment",
+  Custom = "custom",
+}
+
+/**
+ * Gets the configured Azure environment.
+ *
+ * @returns The configured Azure environment from the settings in the built-in authentication provider extension
+ */
+export function getConfiguredAzureEnv(): azureEnv.Environment & { isCustomCloud: boolean } {
+  const authProviderConfig = vscode.workspace.getConfiguration(CustomCloudConfigurationSection);
```

Comment:
I believe some toolkit features do not support non-public azure. Do we need to block logging into non-public azure?